integreat_cms/cms/utils/repair_tree.py

```diff
@@ -0,0 +1,210 @@
+"""
+This module contains utilities to repair or detect inconsistencies in a tree
+"""
+
+from __future__ import annotations
+
+import logging
+from collections import deque
+from typing import TYPE_CHECKING
+
+from django.apps import apps
+from django.db import transaction
+
+from .shadow_instance import ShadowInstance
+
+if TYPE_CHECKING:
+    from typing import Iterable
+
+    from django.apps.registry import Apps
+    from django.db.models import Model
+
+    Page: Model = apps.get_model("cms", "Page")
+
+
+@transaction.atomic
+def repair_tree(
+    page_id: Iterable[int] | None = None,
+    commit: bool = False,
+    logging_name: str = __name__,
+    dj_apps: Apps = apps,
+) -> None:
+    """
+    Fix the tree for a given page, or all trees if no id is given.
+    Changes are only written to the database if ``commit`` is set to ``True``.
+    For details see :class:`MPTTFixer`.
+    """
+    logger = logging.getLogger(logging_name)
+
+    # Use get_model() instead of importing so this function can be used in migrations
+    Page: Model = dj_apps.get_model("cms", "Page")
+    mptt_fixer = MPTTFixer(dj_apps=dj_apps)
+    root_nodes: Iterable[ShadowInstance[Page]]
+
+    if page_id:
+        # Assert that any of the requested pages actually exist
+        for single_id in page_id:
+            try:
+                Page.objects.get(id=single_id)
+            except Page.DoesNotExist as e:
+                raise ValueError(
+                    f'The page with id "{single_id}" does not exist.'
+                ) from e
+        # All ids in page_id are valid, get them from mptt_fixer
+        root_nodes = [
+            mptt_fixer.get_fixed_root_node(single_id) for single_id in page_id
+        ]
+    else:
+        root_nodes = mptt_fixer.get_fixed_root_nodes()
+
+    for root_node in root_nodes:
+        action = "Fixing" if commit else "Detecting problems in"
+        logger.info(
+            "%s tree with id %i... (%r)",
+            action,
+            root_node.tree_id__original,
+            root_node.instance,
+        )
+        for tree_node in mptt_fixer.get_fixed_tree_of_page(root_node.pk):
+            print_changed_fields(tree_node, logging_name=logging_name)
```

Comment:
Instead of printing the entire tree structure, wouldn't it be more useful to just print the changed nodes?

Reply:
In principle yes, but I think then it would just essentially be the same as "*I found something!*" and not much more helpful without further manual investigation of the tree structure in the database. When all nodes of the tree are printed, you have enough context to say "*Yes that makes sense*" or "*Oh now I know why something went wrong in the way it did*"